mono/metadata/icall.c

```diff
@@ -7743,45 +7736,42 @@ mono_ArgIterator_IntGetNextArgType (MonoArgIterator *iter)
 }
 
 MonoObjectHandle
-mono_TypedReference_ToObject (MonoTypedRef* tref, MonoError *error)
+ves_icall_System_TypedReference_ToObject (MonoTypedRef* tref, MonoError *error)
 {
 	return typed_reference_to_object (tref, error);
 }
 
-MonoTypedRef
-mono_TypedReference_MakeTypedReferenceInternal (MonoObject *target, MonoArray *fields)
+void
+ves_icall_System_TypedReference_InternalMakeTypedReference (MonoTypedRef *res, MonoObjectHandle target, MonoArrayHandle fields, MonoReflectionTypeHandle last_field, MonoError *error)
 {
-	MonoTypedRef res;
-	MonoReflectionField *f;
 	MonoClass *klass;
 	MonoType *ftype = NULL;
 	guint8 *p = NULL;
 	int i;
 
-	memset (&res, 0, sizeof (res));
+	memset (res, 0, sizeof (MonoTypedRef));
+
+	g_assert (mono_array_handle_length (fields) > 0);
 
-	g_assert (fields);
-	g_assert (mono_array_length_internal (fields) > 0);
+	klass = mono_handle_class (target);
 
-	klass = target->vtable->klass;
+	for (i = 0; i < mono_array_handle_length (fields); ++i) {
+		MonoClassField *f;
+		MONO_HANDLE_ARRAY_GETVAL (f, fields, MonoClassField*, i);
 
-	for (i = 0; i < mono_array_length_internal (fields); ++i) {
-		f = mono_array_get_internal (fields, MonoReflectionField*, i);
 		g_assert (f);
 
 		if (i == 0)
-			p = (guint8*)target + f->field->offset;
+			p = (guint8*)MONO_HANDLE_RAW (target) + f->offset;
```

Comment:
Using `MONO_HANDLE_RAW` in this loop isn't great.  I think (via asm loading) `mono_class_from_mono_type_internal` can end up invoking managed code, so we don't want to get the address from the handle too early.

How about:
 Change from `guint8 *p` to `intptr_t field_offset`.  In the loop just compute `field_offset` from all the `f->offset`s.
then at the end `res->value = (guint8*)MONO_HANDLE_RAW (target) + field_offset`.

Reply:
Fixed.